create_db.py

```diff
@@ -0,0 +1,265 @@
+		suffixes_fem[key] = suffixes_fem[key]/c
+
+	# Neutral
+	c = 0
+	# loop through keys and create suffixes for each noun. only add suffixes of length less than or equal to 3
+	for nouns in set(seed["Neut"]):
+		suffixes = [nouns[i:] for i in range(len(nouns) - 3, len(nouns))]
+		for s in suffixes:
+			if s in suffixes_neut.keys():
+				suffixes_neut[s] = suffixes_neut[s] + 1
+			else:
+				suffixes_neut[s] = 1
+
+	# delete suffixes with frequency less than 3. Generate probability distribution for rest of the suffixes
+	for key in list(suffixes_neut.keys()):
+		if suffixes_neut[key] < 3:
+			del suffixes_neut[key]
+		else:
+			c = c + suffixes_neut[key]
+
+	for key, val in suffixes_neut.items():
+		suffixes_neut[key] = suffixes_neut[key]/c
+
+
+	return suffixes_masc, suffixes_fem, suffixes_neut
+
+def morphology_based_classification(corpus, seed, suffixes_masc, suffixes_fem, suffixes_neut):
+	# total number of nouns for which there was a calculated gender and a Spacy assigned gender
+	count = 0
+	# nouns where calculated gender and Spacy assigned gender were the same
+	correct = 0
+	# nouns where Spacy did not assign gender
+	spacy_gender_not_exist = 0
+	# total number of nouns in corpus
+	total_nouns = 0
+	for line in corpus:
+	    for idx, token in enumerate(line):
+	    	if token.pos_ == "NOUN":
+	    		# get pronoun for current token
+	    		text = token.text
+	    		# get gold data for gender, to compare against the calculated gender
+	    		spacy_gender = token.morph.get("Gender")
+	    		# total number of nouns in corpus
+	    		total_nouns = total_nouns + 1
+	    		# suffix array for current token calculated to give suffixes with length less than or equal to 3
+	    		suffixes = [text[i:] for i in range(len(text) - 3, len(text))]
+
+	    		# probability of word being neutral/masculine/feminine
+		    	neut_p = 0
+		    	masc_p = 0
+		    	fem_p = 0
+
+		    	for s in suffixes:
+		    		if s in suffixes_neut.keys():
+		    			neut_p = neut_p + suffixes_neut[s]
+
+		    		if s in suffixes_masc.keys():
+		    			masc_p = masc_p + suffixes_masc[s]
+
+		    		if s in suffixes_fem.keys():
+		    			fem_p = fem_p + suffixes_fem[s]
+
+		    	# skip nouns where no suffixes were found
+		    	if neut_p + masc_p + fem_p == 0:
+		    		continue
+
+		    	# normalisation
+		    	masc_p = masc_p/(neut_p + masc_p + fem_p)
+		    	fem_p = fem_p/(neut_p + masc_p + fem_p)
+		    	neut_p = neut_p/(neut_p + masc_p + fem_p)
+
+
+		    	# assign gender based on highest normalised value
+		    	if neut_p > masc_p  and neut_p > fem_p:
+		    		gender = "Neut"
+		    	elif masc_p > neut_p and masc_p > fem_p:
+		    		gender = "Masc"
+		    	elif fem_p > neut_p and fem_p > masc_p:
+		    		gender = "Fem"
+		    	else:
+		    		gender = "tbd"
+
+		    	# check if gender was calculated and spacy assigned gender exists. if it does and the assigned gender is corrected compared
+		    	# to the spacy assignmend, add it to the seed.
+		    	if gender and spacy_gender:
```

Comment:
I am not sure I understand. You use spacy to expand you seed? But spacy is like gold data. So you should not consult them during calculating the gender.

Reply:
agreed and fixed. the explanation is given in the comment below https://github.com/amrtanair/german_noun_gender_predictor-/pull/1#issuecomment-1245620600